Libraries/LibCrypto/Curves/Ed448.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2024, Altomani Gianluca <altomanigianluca@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/ScopeGuard.h>
+#include <LibCrypto/Curves/Ed448.h>
+#include <LibCrypto/OpenSSL.h>
+
+#include <openssl/core_names.h>
+#include <openssl/evp.h>
+
+namespace Crypto::Curves {
+
+ErrorOr<ByteBuffer> Ed448::generate_private_key()
+{
+    auto* key = OPENSSL_TRY_PTR(EVP_PKEY_Q_keygen(nullptr, nullptr, "ED448"));
+    ArmedScopeGuard const free_key { [&] { EVP_PKEY_free(key); } };
+
+    size_t key_size = EVP_PKEY_get_size(key);
+    auto buf = TRY(ByteBuffer::create_uninitialized(key_size));
+
+    OPENSSL_TRY(EVP_PKEY_get_raw_private_key(key, buf.data(), &key_size));
+
+    return buf.slice(0, key_size);
+}
+
+ErrorOr<ByteBuffer> Ed448::generate_public_key(ReadonlyBytes private_key)
+{
+    auto* key = OPENSSL_TRY_PTR(EVP_PKEY_new_raw_private_key(EVP_PKEY_ED448, nullptr, private_key.data(), private_key.size()));
+    ArmedScopeGuard const free_key { [&] { EVP_PKEY_free(key); } };
+
+    size_t key_size = EVP_PKEY_get_size(key);
+    auto buf = TRY(ByteBuffer::create_uninitialized(key_size));
+
+    OPENSSL_TRY(EVP_PKEY_get_raw_public_key(key, buf.data(), &key_size));
+
+    return buf.slice(0, key_size);
+}
+
+ErrorOr<ByteBuffer> Ed448::sign(ReadonlyBytes private_key, ReadonlyBytes message, ReadonlyBytes context)
+{
+    auto* key = OPENSSL_TRY_PTR(EVP_PKEY_new_raw_private_key_ex(nullptr, "ED448", nullptr, private_key.data(), private_key.size()));
+    ArmedScopeGuard const free_key { [&] { EVP_PKEY_free(key); } };
+
+    auto* ctx = EVP_MD_CTX_new();
+    ArmedScopeGuard const free_ctx { [&] { EVP_MD_CTX_free(ctx); } };
+
+    OSSL_PARAM params[] = {
+        OSSL_PARAM_octet_string(OSSL_SIGNATURE_PARAM_CONTEXT_STRING, const_cast<u8*>(context.data()), context.size()),
+        OSSL_PARAM_END
+    };
+
+    OPENSSL_TRY(EVP_DigestSignInit_ex(ctx, nullptr, nullptr, nullptr, nullptr, key, params));
+
+    size_t sig_len = signature_size();
+    auto sig = TRY(ByteBuffer::create_uninitialized(sig_len));
+
+    OPENSSL_TRY(EVP_DigestSign(ctx, sig.data(), &sig_len, message.data(), message.size()));
+
+    return sig.slice(0, sig_len);
+}
+
+ErrorOr<bool> Ed448::verify(ReadonlyBytes public_key, ReadonlyBytes signature, ReadonlyBytes message, ReadonlyBytes context)
+{
+    auto* key = OPENSSL_TRY_PTR(EVP_PKEY_new_raw_public_key_ex(nullptr, "ED448", nullptr, public_key.data(), public_key.size()));
+    ArmedScopeGuard const free_key { [&] { EVP_PKEY_free(key); } };
```

Comment:
instead of using all these scope guards, perhaps we should invent some OpenSSL wrappers?

```c++
class OpenSSLPrivateKey {
public:
    ErrorOr<OpenSSLPrivateKey> create(StringView name, ReadonlyBytes public_key);
    ~OpenSSLPrivateKey();
    
    EVP_PKEY* ptr() { return m_ptr; }
private:
    OpenSSLPrivateKey(EVP_PKEY* ptr) : m_ptr(ptr) {};
    
    EVP_PKEY* m_ptr { nullptr }; 
};
```

Reply:
That makes sense, at least for the most common types, I'll protorype something

Reply:
👍 to me, a one-off scope guard is a handy tool, but using several identical scope guards in every method of a class screams "add some RAII"

Reply:
Really not sure about `TRY(OpenSSL_PKEY::create(EVP_PKEY_Q_keygen, nullptr, nullptr, "ED448"));`, does it make any sense?

Reply:
Yeah that was dumb, should make for sense now